File: 64 Minimum Path Sum/solution.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int> > helper;
        for (int i = 0; i < m; i++) {
            vector<int> level;
            helper.push_back(level);
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0) {
                    helper[i].push_back(grid[i][j]);
                }
                else if (i == 0) {
                    helper[i].push_back(helper[i][j - 1] + grid[i][j]);
                }
                else if (j == 0) {
                    helper[i].push_back(helper[i - 1][j] + grid[i][j]);
                }
                else {
                    helper[i].push_back(min(helper[i - 1][j], helper[i][j - 1]) + grid[i][j]);
                }
            }
        }
        return helper[m - 1][n - 1];
    }
};

File: 64 Minimum Path Sum/solution.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid:
            return 0
        m = len(grid)
        n = len(grid[0])
        helper = [[0 for i in range(n)] for j in range(m)]
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    helper[i][j] = grid[i][j]
                elif i == 0:
                    helper[i][j] = helper[i][j - 1] + grid[i][j]
                elif j == 0:
                    helper[i][j] = helper[i - 1][j] + grid[i][j]
                else:
                    helper[i][j] = min(helper[i][j - 1], helper[i - 1][j]) + grid[i][j]
        return helper[-1][-1]
        